{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Methods for Lineage Reconstruction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Irepan Salvador-Martínez** \n",
    "\n",
    "*Centro Nacional de Análisis Genómico, C/Baldiri Reixac 4, 08028 Barcelona, Spain.*\n",
    "\n",
    "*irepan_salvador@hotmail.com*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Abstract"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The recent development of genetic lineage recorders, designed to register the genealogical history of cells using induced somatic mutations, has opened the possibility of reconstructing complete animal cell lineages. In order to reconstruct a cell lineage tree from a molecular recorder, it is crucial to use an appropriate reconstruction algorithm. \n",
    "Current approaches include algorithms specifically designed for cell lineage reconstruction and the repurposing of phylogenetic algorithms. These methods have, however, the same objective: to uncover the hierarchical relationships between cells and the sequence of cell divisions that have occurred during development. \n",
    "In this chapter, I will use the phylogenetic software FastTree to reconstruct a lineage tree, in a step-by-step manner, using data from a simulated CRISPR-Cas9 recorder. To ensure reproducibility,the code is presented as a Jupyter notebook, available (together with the necessary input files) at https://github.com/irepansalvador/lineage_reconstruction_chapter ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Key words:** Computational methods, Lineage reconstruction, FastTree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In recent years, the development and utilization of genetic lineage recorders have offered a new perspective on studying cell lineages. One of the game-changing tools in this field is CRISPR-based molecular technology. By harnessing the power of CRISPR-Cas9, scientists can introduce targeted genetic modifications into cells, allowing the recording of their lineage relationships through mutations accumulated over time.\n",
    "The process involves the insertion of a synthetic construct into the genetic material of cells, which can accumulate stochastic mutations upon the induction of CRISPR-Cas9 activity during cell differentiation. These mutations serve as a molecular archive, akin to DNA barcodes, which can be read and deciphered through sequencing techniques. By analyzing the patterns of mutations, researchers can reconstruct the cell lineage, shedding light on the developmental paths taken by individual cells.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The explosion of lineage tracing technologies has also highlighted the importance of using appropriate lineage reconstruction methods.\n",
    "In some cases, researchers have tailored algorithms specifically designed for cell lineage reconstruction [1, 2], while others have applied phylogenetic reconstruction algorithms to reconstruct cell lineages [3,4].\n",
    "A recent benchmarking effort conducted via a crowdsourcing innitiative by the DREAM Challenge [5] encouraged participants from diverse fields to submit lineage reconstruction algorithms to reconstruct lineages from different nature and sizes. \n",
    "The best performers from this bechmarking study included a distance-based method (\"DCLEAR\", [1]) and a Machine learning algorithm (\"AMberRland\", [2]). Interestingly, the widely used phylogenetic reconstruction algorithm FastTree [6] (used as a reference algorithm in the study) performed very well, suggesting it is a reliable and effective tool for cell lineage reconstruction."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this chapter, I will use FastTree to reconstruct a cell lineage as I consider it has several advantages that make it a suitable choice for cell lineage reconstruction. \n",
    "Firstly, its speed and efficiency are noteworthy. With the increasing amount of data generated by lineage tracing technologies, it is crucial to have a reconstruction method that can handle large datasets in a reasonable amount of time. In this sense, FastTree's ability to reconstruct phylogenetic trees from thousands or even millions of sequences quickly and accurately makes it a good candidate.\n",
    "Additionally, FastTree uses an approximately maximum-likelihood method, a robust and reliable approach for reconstructing phylogenetic trees. Lastly, an important practical advantage is its ease of installation and use, making it accesible to a wide range of researchers with varying levels of computational expertise."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Requirements"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This chapter, including the example code for lineage reconstruction, is written in a Jupyter Notebook with Python as programming language. Jupyter notebooks are documents that combine computer code, plain text, data, rich visualizations like 3D models, charts, graphs and figures, and interactive controls. They provide a fast interactive environment for exploring and visualizing data that can be shared with others and therefore facilitate code reproducibility. \n",
    "The Jupyter default programming language (called “kernel” in the Jupyter ecosystem) is Python, although many other languages (e.g. Julia, R, Matlab, Octave) are also supported and can be used by installing their specific kernels. For detailed information of Jupyter notebooks visit https://docs.jupyter.org/en/latest/.\n",
    "\n",
    "The specific requirements to run this notebook are:\n",
    "\n",
    "- Jupyter Notebook package\n",
    "- Python packages: Biopython and Dendropy\n",
    "- FastTree software\n",
    "- Input files (FASTA file and transition matrix file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Jupyter Notebook\n",
    "\n",
    "In order to run this notebook it is necessary to install the Jupyter notebook package. In here we will use `pip`, the recommended installation tool for Python. For additional information on the installation of Jupyter please visit the official Jupyter webpage [https://jupyter.org/install](https://jupyter.org/install).\n",
    "\n",
    "To install Jupyter, run the following command in the terminal: \n",
    "\n",
    "`pip install notebook`\n",
    "\n",
    "Once installed we can open Jupyter notebook with the following command:\n",
    "\n",
    "`jupyter notebook`\n",
    "\n",
    "This will open a dashboard in your default browser where you can see the files located in the folder from where you executed the command. If the command is run from the same folder where this notebook is located the dashboard should look similar to the following image."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Jupyter dashboard](\"images/dashboard.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"images/dashboard.png\" alt=\"Jupyter dashboard\" width=\"600\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clicking on the on the Jupyter notebook file (\".ipynb\") will open the notebook on a new browser tab. The notebook can be run cell-by-cell using the interactive buttons on the top bar (see image below). To run a specific cell, first select the cell by clicking on it and then click the \"Run\" button."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"images/notebook.png\" alt=\"Jupyter notebook\" width=\"600\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Python packages: Biopython and Dendropy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After installing the Jupyter notebook, we need to install the following Python packages:\n",
    "\n",
    "- Biopython\n",
    "- DendroPy\n",
    "\n",
    "#### Biopython\n",
    "\n",
    "The Biopython Project [7] is an open-source collection of non-commercial Python tools for computational biology and bioinformatics, created by an international association of developers. With Byopython we can parse the most common bioinformatics file formats into Python utilizable data structures. In here, we will use Biopython to parse a FASTA file into a list with sequences and into another list with cell IDs.\n",
    "Visit the Biopython Tutorial and Cookbook webpage ([http://biopython.org/DIST/docs/tutorial/Tutorial.html](http://biopython.org/DIST/docs/tutorial/Tutorial.html)) to get detailed documentation, including information to get started with Biopython, and specific documentation on a number of modules.\n",
    "\n",
    "\n",
    "To install Biopython run this command in the terminal:\n",
    "\n",
    "`pip install biopython`\n",
    "\n",
    "\n",
    "#### DendroPy \n",
    "\n",
    "DendroPy is a Python library for phylogenetic computing. It provides functions for the simulation, processing, and manipulation of phylogenetic trees and character matrices. It also supports the reading and writing of phylogenetic data in a range of formats, such as NEXUS, NEWICK, Phylip, etc.  \n",
    "In our example we will use Dendropy to visualise the reconstructed lineage tree. It can also be used to compare and summarise trees, to prune tree nodes or to extracte a subtree of interest. A detailed tutorial on how to use the DendroPy library, with annotated practical examples can be found in [https://dendropy.org/primer/](https://dendropy.org/primer/).\n",
    "\n",
    "\n",
    "To install DendroPy run this command in the terminal:\n",
    "\n",
    "`pip install DendroPy`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### FastTree software\n",
    "\n",
    "FastTree is a phylogenetic software that infers approximately-maximum-likelihood phylogenetic trees from alignments of nucleotide or protein sequences. A great advantage of FastTree is that it can handle alignments with up to a million of sequences in a reasonable amount of time and memory.\n",
    "The easiest way to use FastTree it is to download a binary executable file. Ready to execute binary files for the different OS are provided in [www.microbesonline.org/fasttree/#Install](www.microbesonline.org/fasttree/#Install).\n",
    "\n",
    "In this example I am using the Linux 64-bit executable (+SSE) double-precision executable binary file (\"FastTreeDbl\" binary file). The double-precision version is better to resolve very-short branch lengths (i.e. nearly-identical sequences) accurately. For more details on FastTree vistit the webpage [http://www.microbesonline.org/fasttree/](http://www.microbesonline.org/fasttree/) and see the original paper of Price et al [6]."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FASTA file \n",
    "For this example I assume sequencing data from a CRISPR-Cas9 lineage recorder is already available and has already passed the relevant quality controls.\n",
    "To use a phylogenetic reconstruction algorithm like FastTree, the mutations observed in the sequencing reads need to be coded using a character state matrix and formatted in a file that can be recognised by the software. In this example we will use the FASTA format, a text-based format for representing either nucleotide sequences or amino acid (protein) sequences, in which nucleotides or amino acids are represented using single-letter codes.\n",
    "In general, a FASTA file usually contains multiple sequences, where each sequence is preceded by a header line that starts with the \">\" character. This header usually includes the gene ID name, the sequencing read ID, or any other identifier. The next line after the header contains the actual sequence.\n",
    "\n",
    "Using the following command we can see the first 20 lines of the FASTA file we will use as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">unmutated\r\n",
      "000000000000000000000000000000\r\n",
      ">cell_001\r\n",
      "0000000000-------H0100G0000000\r\n",
      ">cell_002\r\n",
      "0000000000-------H0100G0000000\r\n",
      ">cell_003\r\n",
      "000000---------------AG0030000\r\n",
      ">cell_004\r\n",
      "000000---------------AG0030000\r\n",
      ">cell_005\r\n",
      "-H0_0-----B4009000D00000000000\r\n",
      ">cell_006\r\n",
      "-H0_0-----B4009000D0000000D000\r\n",
      ">cell_007\r\n",
      "-H0_00000000009000D00005020000\r\n",
      ">cell_008\r\n",
      "-H0_00000000-----------------D\r\n",
      ">cell_009\r\n",
      "002-----------D000000000-I000D\r\n"
     ]
    }
   ],
   "source": [
    "!head -20 INPUT/05div_30_targets_20_states.fasta "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Transition table\n",
    "\n",
    "A transition (or substitution) matrix describes the frequency at which a character in a nucleotide sequence or a protein sequence changes to other character states over evolutionary time. In case of protein sequences, the scores correspond to a substitution model which that includes also amino-acid stationary frequencies.\n",
    "\n",
    "According to the FastTree webpage, the file to be used as a transition matrix needs to fulfill the following requirements (for further details visit the FastTree webpage):  \n",
    "\n",
    "The transition matrix file must be tab-delimited, with the columns A R N D C Q E G H I L K M F P S T W Y V * (in that order). Each entry gives the rate at which one amino acid changes to another. The last column (\"*\") should have the stationary distribution. Each row must have the row's name as its first field (in order, A R N ... V).\n",
    "The transition matrix Q and the stationary distribution $\\pi$ should satisfy:\n",
    "\n",
    "- Each amino acid is reachable: each element of $\\pi$ must be positive\n",
    "- The frequencies of the amino acids sum to 1: $\\sum \\pi_{i} = 1$.\n",
    "- Each amino acid can mutate away to other amino acids: each element on the diagonal of $Q$ is negative\n",
    "- Each element off the diagonal of $Q$ is non-negative\n",
    "- When at the stationary distribution, the frequencies of each amino acid do not change: $Q · \\pi = 0$\n",
    "- The total frequency of all amino acids does not change as the distribution evolves: the sum of each column M is zero\n",
    "- The average rate of mutation is 1: $\\sum Q_{ii} \\pi_{i} = -1$\n",
    "- Evolution is time reversible: $Q_{ij} / Q_{ji} = \\pi_{i} / s_{j}$ \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In here we will use a simplified transition matrix, where any state can reach another state with the same probability. Following the requierements of FastTree on the transition table, we will use a transition table that contains a `20x20` matrix, with column and row names corresponding to the 20 amino acid (AA) symbols and values of `0.052631579` for all the cells, except for the diagonal that has values of `-1`. An additional column, named `*`, has values of `0.05` for all rows.\n",
    "We can look at the transition table file with the following command (for readability, only the first 50 characters of the first 10 lines are shown):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\tR\tN\tD\tC\tQ\tE\tG\tH\tI\tL\tK\tM\tF\tP\tS\tT\tW\tY\tV\t*\r\n",
      "A\t-1\t0.052631579\t0.052631579\t0.052631579\t0.0526315\r\n",
      "R\t0.052631579\t-1\t0.052631579\t0.052631579\t0.0526315\r\n",
      "N\t0.052631579\t0.052631579\t-1\t0.052631579\t0.0526315\r\n",
      "D\t0.052631579\t0.052631579\t0.052631579\t-1\t0.0526315\r\n",
      "C\t0.052631579\t0.052631579\t0.052631579\t0.052631579\t\r\n",
      "Q\t0.052631579\t0.052631579\t0.052631579\t0.052631579\t\r\n",
      "E\t0.052631579\t0.052631579\t0.052631579\t0.052631579\t\r\n",
      "G\t0.052631579\t0.052631579\t0.052631579\t0.052631579\t\r\n",
      "H\t0.052631579\t0.052631579\t0.052631579\t0.052631579\t\r\n"
     ]
    }
   ],
   "source": [
    "!head -10 INPUT/transition_table.txt | cut -c-50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods\n",
    "\n",
    "In this section I will describe how to reconstruct a lineage tree from sequences coming from a CRISPR-Cas9 recorder. The code used for this example is based on the FastTree reconstruction\n",
    "example provided by Gong et al [5], available at https://github.com/ofirr/dream_examples .\n",
    "The code blocks can be executed if this notebook is downloaded and the software requirements are met (see above). To help with code interpretability, before each code block I describe its objective with a brief text. \n",
    "Additional comments are also included within the code blocks (for example describing the meaning of a value or a parameter). These comments are always preceded by the `#` character.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import packages\n",
    "\n",
    "As a first step, we will import all the necessary packages to run this example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import dendropy        # See the Requirements section to install this package \n",
    "from Bio import SeqIO  # See the Requirements section to install this package"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code block will print the version of packages used in this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>package</th>\n",
       "      <th>jupyterlab</th>\n",
       "      <th>ipython</th>\n",
       "      <th>DendroPy</th>\n",
       "      <th>biopython</th>\n",
       "      <th>pandas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>version</th>\n",
       "      <td>3.0.3</td>\n",
       "      <td>7.19.0</td>\n",
       "      <td>4.6.1</td>\n",
       "      <td>1.78</td>\n",
       "      <td>1.1.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "package jupyterlab ipython DendroPy biopython pandas\n",
       "version      3.0.3  7.19.0    4.6.1      1.78  1.1.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pkg_resources\n",
    "from IPython.display import display\n",
    "root_packages = ['biopython', 'dendropy','ipython', \n",
    "                 'subprocess','jupyterlab', 'pandas']\n",
    "root_packages_list = []\n",
    "for m in pkg_resources.working_set:\n",
    "    if m.project_name.lower() in root_packages:\n",
    "        root_packages_list.append([m.project_name, m.version])\n",
    "display(pd.DataFrame(root_packages_list,columns=[\"package\", \"version\"]\n",
    "                    ).set_index(\"package\").transpose())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions\n",
    "\n",
    "This is a general function to execute a shell command and return its output as a string. We will use it to run FastTree and retrieve the reconstructed tree as a string (in a NEWICK format)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_command(cmd, get_output=False):\n",
    "    \"run command\"\n",
    "    if get_output == False:\n",
    "        process = subprocess.Popen(cmd)\n",
    "        process.wait()\n",
    "    else:\n",
    "        process = subprocess.Popen(\n",
    "            cmd,\n",
    "            stdout=subprocess.PIPE,\n",
    "            stderr=subprocess.PIPE\n",
    "        )\n",
    "        # wait for the process to terminate\n",
    "        stdout, stderr = process.communicate()\n",
    "        stdout = stdout.decode('UTF-8').strip()\n",
    "        stderr = stderr.decode('UTF-8').strip()\n",
    "        return stdout, stderr, process.returncode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input and output\n",
    "Before running FastTree, we will define the INPUT and OUTPUT folders. In this example, the input folder should contain two files (see Requirements section for a description of these files):\n",
    "\n",
    "-  FASTA file with CRISPR-Cas9 recorder sequences\n",
    "-  Transition table (necessary for FastTree)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define folder for input and output\n",
    "input_dir = 'INPUT/'\n",
    "output_dir = \"OUTPUT/\"\n",
    "\n",
    "# Input file (without the \".fasta\")\n",
    "name = '05div_30_targets_20_states'\n",
    "# Define name for output file with the modified sequences using AA symbols\n",
    "fasta_mod_file  = os.path.join(input_dir, (name + '_for_fasttree.fasta'))\n",
    "# Define name for output file with the reconstructed tree (in NEWICK format)\n",
    "newick_output = os.path.join(output_dir, (name + '.newick'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read the input sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will open the FASTA file with the encoded sequences coming from the CRISPR-Cas9 recorder and store the IDs and sequences in the `ids` and `seqs` lists, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "ids  = []\n",
    "seqs = []\n",
    "\n",
    "for seq_record in SeqIO.parse(input_dir + name + \".fasta\", \"fasta\"):\n",
    "    ids.append(seq_record.id)\n",
    "    seqs.append(str(seq_record.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can print the first ten values of `ids` and `seqs` lists we just created to make sure they are correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">unmutated\t000000000000000000000000000000\r\n",
      ">cell_001\t0000000000-------H0100G0000000\r\n",
      ">cell_002\t0000000000-------H0100G0000000\r\n",
      ">cell_003\t000000---------------AG0030000\r\n",
      ">cell_004\t000000---------------AG0030000\r\n",
      ">cell_005\t-H0_0-----B4009000D00000000000\r\n",
      ">cell_006\t-H0_0-----B4009000D0000000D000\r\n",
      ">cell_007\t-H0_00000000009000D00005020000\r\n",
      ">cell_008\t-H0_00000000-----------------D\r\n",
      ">cell_009\t002-----------D000000000-I000D\r\n"
     ]
    }
   ],
   "source": [
    "for i in range(10):\n",
    "     print('>{}\\t{}\\r'.format(ids[i],seqs[i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** This example has been generated with a computer simulation implemented as in Salvador-Martínez et al 2019 [3].\n",
    "Briefly, in this simulation a cell is implemented as a vector of $m=30$ targets. The simulation starts with one cell, with all its targets in an unmutated state. The initial cell then undergoes $d=5$ cell divisions, growing into a population of $N=32$ cells ($N = 2^d$). Following each cell division, each unmutated target can mutate (with a probability $\\mu_d = 0.05$) to one of several possible mutated states. Once a target is mutated, it can no longer change, either to revert to the unmutated state or to transit to a new state.\n",
    "In this example, I have assigned a \"0\" to an umutated CRISPR-Cas9 target, and the numbers \"1-9\" and letters \"A-J\", to the most frequent mutations. If a target is missing due to an inter-target deletion we assign it the character \"-\".\n",
    "The code for the simulations is available in the Github repository [https://github.com/irepansalvador/CRISPR_recorders_sims](https://github.com/irepansalvador/CRISPR_recorders_sims)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Translate characters into amino acid symbols\n",
    "Then we will translate the encoded mutations into amino-acid characters, so the file can be read by FastTree. We will use a Python \"dictionary\" to assign a AA character to the encoded mutations in our FASTA file according to an associative array.\n",
    "First, we create the `symbol_map` dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "symbol_map = {\n",
    "    '0':'A', '1':'R',\n",
    "    '2':'N', '3':'D',\n",
    "    '4':'C', '5':'Q',\n",
    "    '6':'E', '7':'G',\n",
    "    '8':'H', '9':'I',\n",
    "    'A':'L', 'B':'K',\n",
    "    'C':'M', 'D':'F',\n",
    "    'E':'P', 'F':'S',\n",
    "    'G':'T', 'H':'W',\n",
    "    'I':'Y', 'J':'V', # Here I \"merge\" two rare characters as\n",
    "    '_':'V',          # we can only have 20 AA characters \n",
    "    '-':'-', # Intertarget deletions are considered missing information\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we use the dictionary to modify each sequence in the `seqs` list. The new AA encoded sequences are stored in the `seqs_trans` list. Then we print out the first ten sequences of the `seqs_trans` list to visually check that they are correct. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">unmutated\tAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\r\n",
      ">cell_001\tAAAAAAAAAA-------WARAATAAAAAAA\r\n",
      ">cell_002\tAAAAAAAAAA-------WARAATAAAAAAA\r\n",
      ">cell_003\tAAAAAA---------------LTAADAAAA\r\n",
      ">cell_004\tAAAAAA---------------LTAADAAAA\r\n",
      ">cell_005\t-WAVA-----KCAAIAAAFAAAAAAAAAAA\r\n",
      ">cell_006\t-WAVA-----KCAAIAAAFAAAAAAAFAAA\r\n",
      ">cell_007\t-WAVAAAAAAAAAAIAAAFAAAAQANAAAA\r\n",
      ">cell_008\t-WAVAAAAAAAA-----------------F\r\n",
      ">cell_009\tAAN-----------FAAAAAAAAA-YAAAF\r\n"
     ]
    }
   ],
   "source": [
    "seqs_trans = []\n",
    "tmp = []\n",
    "# getting length of list\n",
    "Nseqs = len(seqs)\n",
    "\n",
    "# Iterating the sequences\n",
    "for i in range(Nseqs):\n",
    "    df = pd.DataFrame(list(seqs[i]))\n",
    "    df = df[0].map(symbol_map)\n",
    "    seqs_trans.append(''.join(df.values))\n",
    "\n",
    "for i in range(10):\n",
    "     print('>{}\\t{}\\r'.format(ids[i],seqs_trans[i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we create a new FASTA file with the new AA encoded sequences. This is the file that will be used as input for FastTree later on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing INPUT/05div_30_targets_20_states_for_fasttree.fasta\n"
     ]
    }
   ],
   "source": [
    "with open(fasta_mod_file, 'w') as f:\n",
    "    print(\"writing \" + fasta_mod_file)\n",
    "    for i in range(Nseqs):\n",
    "        f.write('>{}\\r\\n{}\\r\\n'.format(ids[i],seqs_trans[i]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reconstruct the lineage tree and visualise it \n",
    "\n",
    "First, we will create a variable with the path to the executable FastTree file and a function to run FastTree as a shell command and retrieve its output in a string variable. The output should be the reconstructed tree in NEWICK format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasttree_path = \"/home/irepan/software/FastTreeDbl\" # Indicate path to executable file \n",
    "transition_table = input_dir + \"transition_table.txt\"\n",
    "\n",
    "def run_fasttree(newick_output, fasta_mod_file):\n",
    "    run_command([fasttree_path, \n",
    "                 '-pseudo', \n",
    "                 '-trans', transition_table,\n",
    "                 '-out', newick_output, \n",
    "                 fasta_mod_file])\n",
    "    \n",
    "    reconstructed_tree = dendropy.Tree.get_from_path(\n",
    "        newick_output,\"newick\")\n",
    "    return reconstructed_tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we use the `run_fasttree` function we just created to reconstruct the lineage tree using the FASTA file as the input. The output, a tree in the NEWICK format, will be stores in the `rec_tree` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rec_tree = run_fasttree(newick_output, fasta_mod_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To have a proper representation of the tree, is important that we define the \"root\" of the tree using the \"unmutated\" cell included in the FASTA file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Node object at 0x7f1ac7a55040: 'None' (<Taxon 0x7f1ac7a550a0 'unmutated'>)>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# root the output of FastTree at the leaf node representing the unmutated construct\n",
    "rec_tree.reroot_at_node(rec_tree.find_node_with_taxon_label('unmutated'), update_bipartitions=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we can display the reconstructed tree using the `as_ascii_plot()` function of DendroPy. We can define the width of the displayed tree as number of characters. In here we define 80 as the width."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                               /------- cell 015\n",
      "                                                 /-------------+                \n",
      "                                                 |             \\------- cell 016\n",
      "                                                 |                              \n",
      "/------------------------------------------------+             /------- cell 001\n",
      "|                                                |      /------+                \n",
      "|                                                |      |      \\------- cell 002\n",
      "|                                                \\------+                       \n",
      "|                                                       |      /------- cell 003\n",
      "|                                                       \\------+                \n",
      "|                                                              \\------- cell 004\n",
      "|                                                                               \n",
      "|                                                              /------- cell 007\n",
      "|                                                       /------+                \n",
      "|                                                       |      \\------- cell 008\n",
      "|                                  /--------------------+                       \n",
      "+                                  |                    |      /------- cell 005\n",
      "|                                  |                    \\------+                \n",
      "|                                  |                           \\------- cell 006\n",
      "|      /---------------------------+                                            \n",
      "|      |                           |                           /------- cell 013\n",
      "|      |                           |      /--------------------+                \n",
      "|      |                           |      |                    \\------- cell 014\n",
      "|      |                           \\------+                                     \n",
      "|      |                                  |      /--------------------- cell 012\n",
      "|      |                                  \\------+                              \n",
      "|      |                                         |      /-------------- cell 011\n",
      "|      |                                         \\------+                       \n",
      "\\------+                                                |      /------- cell 009\n",
      "       |                                                \\------+                \n",
      "       |                                                       \\------- cell 010\n",
      "       |                                                                        \n",
      "       |                                                       /------- cell 017\n",
      "       |                                                /------+                \n",
      "       |      /-----------------------------------------+      \\------- cell 018\n",
      "       |      |                                         |                       \n",
      "       |      |                                         \\-------------- cell 020\n",
      "       |      |                                                                 \n",
      "       \\------+                                         /-------------- cell 026\n",
      "              |      /----------------------------------+                       \n",
      "              |      |                                  |      /------- cell 019\n",
      "              |      |                                  \\------+                \n",
      "              |      |                                         \\------- cell 025\n",
      "              \\------+                                                          \n",
      "                     |      /------------------------------------------ cell 023\n",
      "                     |      |                                                   \n",
      "                     |      |                           /-------------- cell 024\n",
      "                     \\------+      /--------------------+                       \n",
      "                            |      |                    |      /------- cell 021\n",
      "                            |      |                    \\------+                \n",
      "                            \\------+                           \\------- cell 022\n",
      "                                   |                                            \n",
      "                                   |      /---------------------------- cell 028\n",
      "                                   |      |                                     \n",
      "                                   \\------+                    /------- cell 027\n",
      "                                          |      /-------------+                \n",
      "                                          |      |             \\------- cell 032\n",
      "                                          \\------+                              \n",
      "                                                 |             /------- cell 029\n",
      "                                                 |      /------+                \n",
      "                                                 \\------+      \\------- cell 030\n",
      "                                                        |                       \n",
      "                                                        \\-------------- cell 031\n",
      "                                                                                \n",
      "                                                                                \n"
     ]
    }
   ],
   "source": [
    "# Alternatively we can also display the branch lengths by uncommenting the following command \n",
    "#print(rec_tree.as_ascii_plot(width=80, plot_metric='length'))\n",
    "\n",
    "print(rec_tree.as_ascii_plot(width=80))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pruning a tree\n",
    "\n",
    "In case you are interested in looking at the reconstructed tree using only a subset of cells, you can \"prune\" the reconstructed tree using the cell IDs using the function `extract_tree_with_taxa_labels()` of Dendropy. This option is especially useful when handling trees with hundreds or thousands of cells.\n",
    "In the following example, we show the topology of a subset tree containing only the first 12 cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        /---------- cell 001\n",
      "                              /---------+                   \n",
      "                              |         \\---------- cell 002\n",
      "/-----------------------------+                             \n",
      "|                             |         /---------- cell 003\n",
      "|                             \\---------+                   \n",
      "|                                       \\---------- cell 004\n",
      "|                                                           \n",
      "+                                       /---------- cell 007\n",
      "|                             /---------+                   \n",
      "|                             |         \\---------- cell 008\n",
      "|         /-------------------+                             \n",
      "|         |                   |         /---------- cell 005\n",
      "|         |                   \\---------+                   \n",
      "\\---------+                             \\---------- cell 006\n",
      "          |                                                 \n",
      "          |         /------------------------------ cell 012\n",
      "          \\---------+                                       \n",
      "                    |         /-------------------- cell 011\n",
      "                    \\---------+                             \n",
      "                              |         /---------- cell 009\n",
      "                              \\---------+                   \n",
      "                                        \\---------- cell 010\n",
      "                                                            \n",
      "                                                            \n"
     ]
    }
   ],
   "source": [
    "# Cell IDs for pruning the tree\n",
    "cells_for_subset_tree = [\"cell 001\",\"cell 002\",\"cell 003\",\"cell 004\",\n",
    "                         \"cell 005\",\"cell 006\",\"cell 007\",\"cell 008\",\n",
    "                         \"cell 009\",\"cell 010\",\"cell 011\",\"cell 012\",\n",
    "                        ]\n",
    "# tree1 is a subset of the reconstructed tree containing only the first 12 cells\n",
    "tree1 = rec_tree.extract_tree_with_taxa_labels(labels=cells_for_subset_tree)\n",
    "\n",
    "# print the subtree topology\n",
    "print(tree1.as_ascii_plot(width=60))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare with reference tree\n",
    "\n",
    "Although is not always possible, in case of having a ground truth tree (or \"reference tree\") we could quantify the accuracy of our lineage reconstruction using Dendropy. As in our example we are using a simulated tree, we have this possibility. A simple metric that can be used to compare two phylogenetic trees is the Robinson–Foulds distance. In here we will use a scaled implementation of the RF distance, so the RF distance can have minimum value of 0 (of the reconstructed tree is identical to the reference tree) and a maximum value of 1 if none of the bifurcations of the reference tree is found in the reconstructed tree). This provides a measure of the global accuracy of the inferred lineage tree. \n",
    "\n",
    "The first step is to read the reference tree file, available as a newick file, creating a Dendropy tree object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "REF_tree = dendropy.Tree.get(\n",
    "    path=os.path.join(input_dir,'{}_REF.nw'.format(name)),\n",
    "    schema=\"newick\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A Dendropy requirement to compare trees is that the trees share a common \"TaxonNamespace\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# establish common taxon namespace\n",
    "tns = dendropy.TaxonNamespace()\n",
    "\n",
    "# ensure all trees loaded use common namespace\n",
    "tree1 = dendropy.Tree.get(\n",
    "        data=REF_tree.as_string(schema=\"newick\"),\n",
    "        schema='newick', taxon_namespace=tns)\n",
    "tree2 = dendropy.Tree.get(\n",
    "        data=rec_tree.as_string(schema=\"newick\"),\n",
    "        schema='newick', taxon_namespace=tns)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we calculate the number of splits (or bipartitions) that difer between the reference and the reconstructed tree using the `dendropy.calculate.treecompare.find_missing_bipartitions()` function. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# splits in REF tree missing in reconstructed tree\n",
    "len(dendropy.calculate.treecompare.find_missing_bipartitions(tree1, tree2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We finally divide the number of wrong splits between the total number of splits of the reference tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.3064516129032258"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dendropy.calculate import treecompare\n",
    "\n",
    "#dendropy.calculate.treecompare.find_missing_bipartitions(tree1, tree2)\n",
    "split_diff = len(dendropy.calculate.treecompare.find_missing_bipartitions(tree1, tree2))\n",
    "all_splits = len(tree1.encode_bipartitions())\n",
    "split_diff / all_splits"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our result is `0.306`, from which we can conclude that 30% of the bifurcations in the reference tree were not found in the reconstructed tree. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "\n",
    "In conclusion, the use of FastTree as the reconstruction algorithm in this study is justified by its speed, accuracy, and ease of use. Its ability to handle large datasets, estimate evolutionary relationships, and produce reliable results make it a good choice for analyzing the genetic lineage recorders generated through CRISPR-based molecular technology."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, it is important to highlight that currently there is no consensus on a reconstruction method and further benchmarking of existent and new reconstruction methods is still necessary.\n",
    "As it was shown in the DREAM challenge, the performance of the reconstruction algorithms depended on the nature and size of the lineage trees to be reconstructed. It is therefore recommended that researchers consider choosing an appropriate reconstruction method based on the features (size, depth, number of targets) of the lineage tree they want to reconstruct."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note about the OS\n",
    "\n",
    "This notebook has been created with a computer with Ubuntu OS (version 20.04.6) and using Python version 3.8.5, as it can be seen from the output of the following commands:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No LSB modules are available.\r\n",
      "Distributor ID:\tUbuntu\r\n",
      "Description:\tUbuntu 20.04.6 LTS\r\n",
      "Release:\t20.04\r\n",
      "Codename:\tfocal\r\n"
     ]
    }
   ],
   "source": [
    "!lsb_release -a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.8.5\r\n"
     ]
    }
   ],
   "source": [
    "!python --version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Ubuntu, as in many other Linux distributions, Python comes preinstalled. Instructions to install Python in the different OS (Windows, Mac, Linux), as well as detailed documentation, tutorials, and guides can be found on the official webpage [https://www.python.org/](https://www.python.org/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "[1] Wuming Gong, Hyunwoo J Kim, Daniel J Garry, and Il-Youp Kwak. Single cell lineage recon-\n",
    "struction using distance-based algorithms and the R package, DCLEAR. BMC Bioinformatics,\n",
    "23(1):103, 2022. ISSN 1471-2105. doi: 10.1186/s12859-022-04633-x.\n",
    "\n",
    "\n",
    "[2] Alisa Prusokiene, Augustinas Prusokas, and Renata Retkute. Machine learning based lineage\n",
    "tree reconstruction improved with knowledge of higher level relationships between cells and\n",
    "genomic barcodes. NAR Genomics and Bioinformatics, 5(3):lqad077, 08 2023. ISSN 2631-9268.\n",
    "doi: 10.1093/nargab/lqad077. URL https://doi.org/10.1093/nargab/lqad077 .\n",
    "\n",
    "\n",
    "[3] Irepan Salvador-Martínez, Marco Grillo, Michalis Averof, and Maximilian J Telford. Is it pos-\n",
    "sible to reconstruct an accurate cell lineage using CRISPR recorders? eLife, 8, jan 2019. ISSN\n",
    "2050-084X. doi: 10.7554/eLife.40292.\n",
    "\n",
    "\n",
    "[4] Aaron McKenna, Gregory M Findlay, J. A. Gagnon, M. S. Horwitz, A. F. Schier, and J. Shen-\n",
    "dure. Whole-organism lineage tracing by combinatorial and cumulative genome editing. Sci-\n",
    "ence, 353(6298):aaf7907–aaf7907, jul 2016. ISSN 0036-8075. doi: 10.1126/science.aaf7907.\n",
    "\n",
    "\n",
    "[5] Wuming Gong, Alejandro A. Granados, Jingyuan Hu, Matthew G. Jones, Ofir Raz, Irepan\n",
    "Salvador-Martínez, Hanrui Zhang, Ke-Huan K. Chow, Il-Youp Kwak, Renata Retkute, Alisa\n",
    "Prusokiene, Augustinas Prusokas, Alex Khodaverdian, Richard Zhang, Suhas Rao, Robert\n",
    "Wang, Phil Rennert, Vangala G. Saipradeep, Naveen Sivadasan, Aditya Rao, Thomas Joseph,\n",
    "Rajgopal Srinivasan, Jiajie Peng, Lu Han, Xuequn Shang, Daniel J. Garry, Thomas Yu, Verena\n",
    "Chung, Michael Mason, Zhandong Liu, Yuanfang Guan, Nir Yosef, Jay Shendure, Maximil-\n",
    "ian J. Telford, Ehud Shapiro, Michael B. Elowitz, and Pablo Meyer. Benchmarked approaches\n",
    "for reconstruction of in vitro cell lineages and in silico models of C. elegans and M. mus-\n",
    "culus developmental trees. Cell Systems, 12(8):810–826.e4, aug 2021. ISSN 24054712. doi:\n",
    "10.1016/j.cels.2021.05.008.\n",
    "\n",
    "[6] Morgan N Price, Paramvir S Dehal, and Adam P Arkin. FastTree 2 - Approximately maximum-\n",
    "likelihood trees for large alignments. PLoS ONE, 5(3):e9490, mar 2010. ISSN 19326203. doi:\n",
    "10.1371/journal.pone.0009490.\n",
    "\n",
    "[7] Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman, Cymon J. Cox, Andrew\n",
    "Dalke, Iddo Friedberg, Thomas Hamelryck, Frank Kauff, Bartek Wilczynski, and Michiel\n",
    "J. L. de Hoon. Biopython: freely available Python tools for computational molecular bi-\n",
    "ology and bioinformatics. Bioinformatics, 25(11):1422–1423, 03 2009. ISSN 1367-4803. doi:\n",
    "10.1093/bioinformatics/btp163."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lineage_r",
   "language": "python",
   "name": "lineage_r"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
